test/tests/filter.js

```diff
@@ -0,0 +1,833 @@
+var assert = require("assert"),
+  promisify = require("promisify-node"),
+  fse = promisify(require("fs-extra")),
+  path = require("path"),
+  local = path.join.bind(path, __dirname);
+
+describe("Filter", function() {
+  var NodeGit = require("../../");
+
+  var emptyRepoPath = local("../repos/empty");
+  var filterName = "psuedo_filter";
+  var Registry = NodeGit.FilterRegistry;
+  var Checkout = NodeGit.Checkout;
+  var Repository = NodeGit.Repository;
+  var reposPath = local("../repos/workdir");
+
+  var packageJsonPath = path.join(reposPath, "package.json");
+  var readmePath = path.join(reposPath, "README.md");
+  
+  var GIT_PASSTHROUGH = -30;
+
+  var mockFilter = {
+    apply: function() {
+      return;
+    },
+    check: function(){
+      return;
+    }
+  };
+
+  function commitFile(repo, fileName, fileContent, commitMessage) {
+    let index;
+    let treeOid;
+    let parent;
+
+    return fse.writeFile(path.join(repo.workdir(), fileName), fileContent)
+    .then(function() {
+      return repo.refreshIndex();
+    })
+    .then(function(indexResult) {
+      index = indexResult;
+    })
+    .then(function() {
+      return index.addByPath(fileName);
+    })
+    .then(function() {
+      return index.write();
+    })
+    .then(function() {
+      return index.writeTree();
+    })
+    .then(function(oidResult) {
+      treeOid = oidResult;
+      return NodeGit.Reference.nameToId(repo, "HEAD");
+    })
+    .then(function(head) {
+      return repo.getCommit(head);
+    })
+    .then(function(parentResult) {
+      parent = parentResult;
+      return Promise.all([
+        NodeGit.Signature.create("Foo Bar", "foo@bar.com", 123456789, 60),
+        NodeGit.Signature.create("Foo A Bar", "foo@bar.com", 987654321, 90)
+      ]);
+    })
+    .then(function(signatures) {
+      let author = signatures[0];
+      let committer = signatures[1];
+
+      return repo.createCommit(
+        "HEAD",
+        author,
+        committer,
+        commitMessage,
+        treeOid,
+        [parent]);
+    });
+  }
+
+  beforeEach(function() {
+    var test = this;
+
+    return Repository.open(reposPath)
+      .then(function(repository) {
+        test.repository = repository;
+        return Repository.open(emptyRepoPath);
+      })
+      .then(function(emptyRepo) {
+        test.emptyRepo = emptyRepo;
+        return fse.writeFile(
+          path.join(reposPath, ".gitattributes"), 
+          "*.md filter="+ filterName +" -text", {
```

Comment:
Spaces between concat operator, please. And if you're going to separate the arguments onto their own lines, don't open the object on the second line, instead:

```javascript
return fse.writeFile(
  path.join(),
  "",
  { encoding: "utf-8" }
);
```